File: kotlin/provenance-mark/src/main/kotlin/com/blockchaincommons/provenancemark/ProvenanceMarkException.kt
package com.blockchaincommons.provenancemark

sealed class ProvenanceMarkException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class InvalidSeedLength(actual: Int) :
        ProvenanceMarkException("invalid seed length: expected 32 bytes, got $actual bytes")

    class DuplicateKey(key: String) :
        ProvenanceMarkException("duplicate key: $key")

    class MissingKey(key: String) :
        ProvenanceMarkException("missing key: $key")

    class InvalidKey(key: String) :
        ProvenanceMarkException("invalid key: $key")

    class ExtraKeys(expected: Int, actual: Int) :
        ProvenanceMarkException("wrong number of keys: expected $expected, got $actual")

    class InvalidKeyLength(expected: Int, actual: Int) :
        ProvenanceMarkException("invalid key length: expected $expected, got $actual")

    class InvalidNextKeyLength(expected: Int, actual: Int) :
        ProvenanceMarkException("invalid next key length: expected $expected, got $actual")

    class InvalidChainIdLength(expected: Int, actual: Int) :
        ProvenanceMarkException("invalid chain ID length: expected $expected, got $actual")

    class InvalidMessageLength(expected: Int, actual: Int) :
        ProvenanceMarkException("invalid message length: expected at least $expected, got $actual")

    class InvalidInfoCbor :
        ProvenanceMarkException("invalid CBOR data in info field")

    class DateOutOfRange(details: String) :
        ProvenanceMarkException("date out of range: $details")

    class InvalidDate(details: String) :
        ProvenanceMarkException("invalid date: $details")

    class MissingUrlParameter(parameter: String) :
        ProvenanceMarkException("missing required URL parameter: $parameter")

    class YearOutOfRange(year: Int) :
        ProvenanceMarkException("year out of range for 2-byte serialization: must be between 2023-2150, got $year")

    class InvalidMonthOrDay(year: Int, month: Int, day: Int) :
        ProvenanceMarkException("invalid month ($month) or day ($day) for year $year")

    class ResolutionError(details: String) :
        ProvenanceMarkException("resolution serialization error: $details")

    class Bytewords(details: String) :
        ProvenanceMarkException("bytewords error: $details")

    class Cbor(details: String) :
        ProvenanceMarkException("CBOR error: $details")

    class Url(details: String) :
        ProvenanceMarkException("URL parsing error: $details")

    class Base64(details: String) :
        ProvenanceMarkException("base64 decoding error: $details")

    class Json(details: String) :
        ProvenanceMarkException("JSON error: $details")

    class IntegerConversion(details: String) :
        ProvenanceMarkException("integer conversion error: $details")

    class Envelope(details: String) :
        ProvenanceMarkException("envelope error: $details")

    class Validation(val issue: ValidationIssue) :
        ProvenanceMarkException("validation error: $issue")
}

typealias Result<T> = kotlin.Result<T>

File: csharp/ProvenanceMark/ProvenanceMark/ProvenanceMarkException.cs
namespace BlockchainCommons.ProvenanceMark;

/// <summary>
/// Base exception type for provenance-mark failures.
/// </summary>
public class ProvenanceMarkException : Exception
{
    public ProvenanceMarkException(string message, Exception? innerException = null)
        : base(message, innerException)
    {
    }

    public static ProvenanceMarkException InvalidSeedLength(int actual) =>
        new($"invalid seed length: expected 32 bytes, got {actual} bytes");

    public static ProvenanceMarkException DuplicateKey(string key) =>
        new($"duplicate key: {key}");

    public static ProvenanceMarkException MissingKey(string key) =>
        new($"missing key: {key}");

    public static ProvenanceMarkException InvalidKey(string key) =>
        new($"invalid key: {key}");

    public static ProvenanceMarkException ExtraKeys(int expected, int actual) =>
        new($"wrong number of keys: expected {expected}, got {actual}");

    public static ProvenanceMarkException InvalidKeyLength(int expected, int actual) =>
        new($"invalid key length: expected {expected}, got {actual}");

    public static ProvenanceMarkException InvalidNextKeyLength(int expected, int actual) =>
        new($"invalid next key length: expected {expected}, got {actual}");

    public static ProvenanceMarkException InvalidChainIdLength(int expected, int actual) =>
        new($"invalid chain ID length: expected {expected}, got {actual}");

    public static ProvenanceMarkException InvalidMessageLength(int expected, int actual) =>
        new($"invalid message length: expected at least {expected}, got {actual}");

    public static ProvenanceMarkException InvalidInfoCbor() =>
        new("invalid CBOR data in info field");

    public static ProvenanceMarkException DateOutOfRange(string details) =>
        new($"date out of range: {details}");

    public static ProvenanceMarkException InvalidDate(string details) =>
        new($"invalid date: {details}");

    public static ProvenanceMarkException MissingUrlParameter(string parameter) =>
        new($"missing required URL parameter: {parameter}");

    public static ProvenanceMarkException YearOutOfRange(int year) =>
        new($"year out of range for 2-byte serialization: must be between 2023-2150, got {year}");

    public static ProvenanceMarkException InvalidMonthOrDay(int year, int month, int day) =>
        new($"invalid month ({month}) or day ({day}) for year {year}");

    public static ProvenanceMarkException ResolutionError(string details) =>
        new($"resolution serialization error: {details}");

    public static ProvenanceMarkException Bytewords(string details, Exception? innerException = null) =>
        new($"bytewords error: {details}", innerException);

    public static ProvenanceMarkException Cbor(string details, Exception? innerException = null) =>
        new($"CBOR error: {details}", innerException);

    public static ProvenanceMarkException Url(string details, Exception? innerException = null) =>
        new($"URL parsing error: {details}", innerException);

    public static ProvenanceMarkException Base64(string details, Exception? innerException = null) =>
        new($"base64 decoding error: {details}", innerException);

    public static ProvenanceMarkException Json(string details, Exception? innerException = null) =>
        new($"JSON error: {details}", innerException);

    public static ProvenanceMarkException IntegerConversion(string details) =>
        new($"integer conversion error: {details}");

    public static ProvenanceMarkException Envelope(string details, Exception? innerException = null) =>
        new($"envelope error: {details}", innerException);
}

/// <summary>
/// Validation-specific exception that preserves the flagged issue.
/// </summary>
public sealed class ProvenanceMarkValidationException : ProvenanceMarkException
{
    public ProvenanceMarkValidationException(ValidationIssue issue)
        : base($"validation error: {issue}")
    {
        Issue = issue;
    }

    public ValidationIssue Issue { get; }
}
